formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt

```diff
@@ -0,0 +1,90 @@
+package kotlinx.serialization.json.internal
+
+/*
+ * Optimized version of StringBuilder that is specific to JSON-encoding
+ */
+internal actual class JsonStringBuilder {
+    private var array = CharArrayPool.take()
+    private var size = 0
+
+    actual fun append(value: Long) {
+        ensureAdditionalCapacity(20) // Long length
+        append(value.toString())
+    }
+
+    actual fun append(ch: Char) {
+        ensureAdditionalCapacity(1)
+        array[size++] = ch
+    }
+
+    actual fun append(string: String) {
+        val length = string.length
+        ensureAdditionalCapacity(length)
+        string.toCharArray(array, size, 0, string.length)
+        size += length
+    }
+
+    actual fun appendQuoted(string: String) {
+        ensureAdditionalCapacity(string.length + 2)
+        val arr = array
+        var sz = size
+        arr[sz++] = '"'
+        val length = string.length
+        string.toCharArray(arr, sz, 0, length)
+        for (i in sz until sz + length) {
+            val ch = arr[i].toInt()
+            // Do we have unescaped symbols?
+            if (ch < ESCAPE_MARKERS.size && ESCAPE_MARKERS[ch] != 0.toChar()) {
+                // Go to slow path
+                return appendStringSlowPath(i - sz, i, string)
+            }
+        }
+        // Update the state
+        sz += length
+        arr[sz++] = '"'
+        size = sz
+    }
+
+    private fun appendStringSlowPath(firstEscapedChar: Int, currentSize: Int, string: String) {
+        var sz = currentSize
+        for (i in firstEscapedChar until string.length) {
+            val ch = string[i].toInt()
+            // Do we have unescaped symbols?
+            var marker: Char = 0.toChar()
+            if (ch < ESCAPE_MARKERS.size && ESCAPE_MARKERS[ch].also { marker = it } != 0.toChar()) {
+                if (marker != 0.toChar()) {
+                    array[sz] = '\\'
+                    array[sz + 1] = marker
+                    sz += 2
+                } else {
+                    val escapedString = ESCAPE_STRINGS[ch]!!
```

Comment:
I don't get it. How we can visit this branch if marker seems to be always non-zero? (Given that this all is under `ESCAPE_MARKERS[ch].also { marker = it } != 0.toChar()` branch).

Also, how we distinguish regular chars (say 'A', which is 0x41) from 0..0x1f chars? They all have `0` value in `ESCAPE_MARKERS` array.

Reply:
I feel that you should tests for some corner cases, benchmarks are not enough. For example, test with vertical tab (0x0b) or something and make sure behavior stays the same.

Reply:
Added tests, changed code layout, verified our benchmarks one more time